src/services/bunker.py

```diff
@@ -1,15 +1,344 @@
+                for e, e_vals in per_epoch_buckets.items() if min_bounded_epoch <= e <= epoch
+            ])  # aka state.slashings from spec
+            adjusted_total_slashing_balance = min(
+                bounded_slashings_balance_sum * PROPORTIONAL_SLASHING_MULTIPLIER_BELLATRIX,
+                total_balance
+            )
+            for v in lido_validators_slashed_in_epoch:
+                effective_balance = int(v.validator.effective_balance)
+                penalty_numerator = effective_balance // EFFECTIVE_BALANCE_INCREMENT * adjusted_total_slashing_balance
+                penalty = penalty_numerator // total_balance * EFFECTIVE_BALANCE_INCREMENT
+                per_epoch_lido_midterm_penalties[epoch] += penalty
+
+        min_lido_midterm_penalty_epoch = min(per_epoch_lido_midterm_penalties.keys())
+
+        # Put per epoch buckets into per frame buckets to calculate lido midterm penalties impact in each frame
+        per_frame_buckets = defaultdict(int)
+        left_border = min_lido_midterm_penalty_epoch
+        right_border = self.f_conf.initial_epoch + self.f_conf.epochs_per_frame
+        # The first bucket should have flexible size
+        while right_border <= min_lido_midterm_penalty_epoch:
+            right_border += self.f_conf.epochs_per_frame
+        index = 0
+        for epoch, midterm_penalty_sum in sorted(per_epoch_lido_midterm_penalties.items()):
+            if epoch > right_border:
+                index += 1
+                left_border = right_border + 1
+                right_border += self.f_conf.epochs_per_frame
+            if left_border <= epoch <= right_border:
+                per_frame_buckets[index] += midterm_penalty_sum
+
+        # If any midterm penalty sum of lido validators in frame bucket greater than rebase we should trigger bunker
+        max_lido_midterm_penalty = max(per_frame_buckets.values())
+        logger.info({"msg": f"Max lido midterm penalty: {max_lido_midterm_penalty}"})
+        if max_lido_midterm_penalty > cl_rebase:
+            return True
+
+        return False
+
+    def _is_abnormal_cl_rebase(self, cdtp: CommonDataToProcess, frame_cl_rebase: Gwei) -> bool:
+        logger.info({"msg": "Checking abnormal CL rebase"})
+        normal_cl_rebase = self._calculate_normal_cl_rebase(cdtp)
+        if frame_cl_rebase < normal_cl_rebase:
+            logger.info({"msg": "Abnormal CL rebase in frame"})
+            nearest_cl_rebase, far_cl_rebase = self._calculate_nearest_and_far_cl_rebase(cdtp)
+            if nearest_cl_rebase < 0 or far_cl_rebase < 0:
+                return True
+        return False
+
+    def _calculate_normal_cl_rebase(self, cdtp: CommonDataToProcess) -> Gwei:
+        """
+        Calculate normal CL rebase (relative to all validators and the previous Lido frame)
+        for current frame for Lido validators
+        """
+        last_report_blockstamp = self.cm.get_first_non_missed_slot(cdtp.ref_blockstamp, cdtp.last_report_ref_slot)
```

Comment:
Lets move this method outside conensus class and make it static

Reply:
done